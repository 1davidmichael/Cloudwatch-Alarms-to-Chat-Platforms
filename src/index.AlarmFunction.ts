import * as axios from 'axios';


exports.handler = async function(event: any, _context: any) {
  console.log('event:', JSON.stringify(event));

  const webhookUri = process.env.MS_TEAMS_WEBHOOK as string;

  const alarm = JSON.parse(event.Records[0].Sns.Message);
  const alarmName = alarm.AlarmName;
  const alarmState = alarm.NewStateValue;
  const alarmReason = alarm.NewStateReason;

  const alarmColor = getColor(alarmState);

  console.log(`${alarmName} - ${alarmReason} - ${alarmState} - ${alarmColor}`);


  const webhookBody = JSON.stringify({
    '@type': 'MessageCard',
    '@context': 'https://schema.org/extensions',
    'summary': alarmName,
    'themeColor': alarmColor,
    'title': 'CloudWatch Notification',
    'sections': [
      {
        activityTitle: alarmName,
        activitySubtitle: alarmState,
        activityImage: 'https://upload.wikimedia.org/wikipedia/commons/thumb/1/15/AWS_Simple_Icons_Monitoring_Amazon_CloudWatch.svg/1200px-AWS_Simple_Icons_Monitoring_Amazon_CloudWatch.svg.png',
        text: alarmReason,
      },
    ],
  });

  const response = await axios.default.post(webhookUri, webhookBody);
  console.log(`Response is: ${response.status}`);

};

function getColor(alarmState: string): string {

  let alarmColor = 'FFC300';

  if ( alarmState == 'ALARM') {
    alarmColor = 'FF0000';
  } else if ( alarmState == 'OK') {
    alarmColor = '27AE60';
  }

  return alarmColor;
}