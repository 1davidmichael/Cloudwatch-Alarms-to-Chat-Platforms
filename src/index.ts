import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as cw_actions from '@aws-cdk/aws-cloudwatch-actions';
import { Runtime } from '@aws-cdk/aws-lambda';
import { SnsEventSource } from '@aws-cdk/aws-lambda-event-sources';
import * as lambda from '@aws-cdk/aws-lambda-nodejs';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';

export interface CloudwatchAlarmsToTeamsConstructProps {
  /**
   * Provide a webhook url.
   *
   */
  readonly webhookUrl: string;
}

export class CloudwatchAlarmsToTeamsConstruct extends cdk.Construct {

  public readonly lambdaFunction: lambda.NodejsFunction;
  public readonly topic: sns.Topic;

  constructor(scope: cdk.Construct, id: string, props: CloudwatchAlarmsToTeamsConstructProps) {
    super(scope, id);

    this.topic = new sns.Topic(this, 'SNSTopic');

    this.lambdaFunction = new lambda.NodejsFunction(this, 'AlarmFunction', {
      runtime: Runtime.NODEJS_14_X,
      handler: 'handler',
      environment: {
        MS_TEAMS_WEBHOOK: props.webhookUrl,
      },
    });

    this.lambdaFunction.addEventSource(new SnsEventSource(this.topic));
  }

  public addAlarmToTeamsNotification(alarm: cloudwatch.Alarm) {
    alarm.addOkAction(new cw_actions.SnsAction(this.topic));
    alarm.addAlarmAction(new cw_actions.SnsAction(this.topic));
  }
}
