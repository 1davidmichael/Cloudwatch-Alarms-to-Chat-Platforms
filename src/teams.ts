import * as path from 'path';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as cw_actions from '@aws-cdk/aws-cloudwatch-actions';
import { Runtime } from '@aws-cdk/aws-lambda';
import { SnsEventSource } from '@aws-cdk/aws-lambda-event-sources';
import * as lambda from '@aws-cdk/aws-lambda-python';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';


export interface CloudwatchAlarmsToTeamsProps {
  /**
   * Provide a webhook url.
   *
   */
  readonly webhookUrl: string;
}

export class CloudwatchAlarmsToTeams extends cdk.Construct {

  public readonly lambdaFunction: lambda.PythonFunction;
  public readonly topic: sns.Topic;

  constructor(scope: cdk.Construct, id: string, props: CloudwatchAlarmsToTeamsProps) {
    super(scope, id);

    this.topic = new sns.Topic(this, 'SNSTopic');

    this.lambdaFunction = new lambda.PythonFunction(this, 'AlarmFunction', {
      runtime: Runtime.PYTHON_3_8,
      entry: path.join(__dirname, '../functions', 'teams'),
      description: 'CloudWatch Alarms to Microsoft Teams Webhook',
      environment: {
        MS_TEAMS_WEBHOOK: props.webhookUrl,
      },
    });

    this.lambdaFunction.addEventSource(new SnsEventSource(this.topic));
  }

  public addAlarmToTeamsNotification(alarm: cloudwatch.Alarm) {
    alarm.addOkAction(new cw_actions.SnsAction(this.topic));
    alarm.addAlarmAction(new cw_actions.SnsAction(this.topic));
  }
}
