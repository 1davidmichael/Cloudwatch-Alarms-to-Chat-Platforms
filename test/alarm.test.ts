import { expect as expectCDK, countResources, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { CloudwatchAlarmsToTeams } from '../src';

/*
 * Example test
 */
test('SNS Topic Created', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');
  // WHEN
  new CloudwatchAlarmsToTeams(stack, 'MyTestConstruct', {
    webhookUrl: 'https://testhook.com',
  });
  // THEN
  expectCDK(stack).to(countResources('AWS::SNS::Topic', 1));
});

test('Lambda Function Created', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');
  // WHEN
  new CloudwatchAlarmsToTeams(stack, 'MyTestConstruct', {
    webhookUrl: 'https://testhook.com',
  });
  // THEN
  expectCDK(stack).to(countResources('AWS::Lambda::Function', 1));
});

test('Snapshot Test', () => {
  const stack = new cdk.Stack();
  new CloudwatchAlarmsToTeams(stack, 'Test', {
    webhookUrl: 'https://testhook.com',
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
