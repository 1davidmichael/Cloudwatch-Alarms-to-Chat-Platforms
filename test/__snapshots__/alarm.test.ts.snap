// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9ArtifactHashCB6B738D": Object {
      "Description": "Artifact hash for asset \\"d03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9\\"",
      "Type": "String",
    },
    "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9S3BucketFA1E25A2": Object {
      "Description": "S3 bucket for asset \\"d03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9\\"",
      "Type": "String",
    },
    "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9S3VersionKey9D308110": Object {
      "Description": "S3 key for asset version \\"d03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestAlarmFunction0CE806B6": Object {
      "DependsOn": Array [
        "TestAlarmFunctionServiceRoleBB959AB8",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9S3BucketFA1E25A2",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9S3VersionKey9D308110",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd03cd3580f0b71052a95d156ee5f2ced579b589d1ff7f89500204c40aea5fbe9S3VersionKey9D308110",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "MS_TEAMS_WEBHOOK": "https://testhook.com",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestAlarmFunctionServiceRoleBB959AB8",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestAlarmFunctionAllowInvokeTestSNSTopicC77C5CD20A249902": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TestAlarmFunction0CE806B6",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "TestSNSTopicAFADF243",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestAlarmFunctionSNSTopic72DD8974": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "TestAlarmFunction0CE806B6",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "TestSNSTopicAFADF243",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestAlarmFunctionServiceRoleBB959AB8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestSNSTopicAFADF243": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
